diff --git a/system/kubeovn/install.sh b/system/kubeovn/install.sh
index 6120c35..29a3146 100644
--- a/system/kubeovn/install.sh
+++ b/system/kubeovn/install.sh
@@ -49,9 +49,9 @@ REGISTRY="docker.io/kubeovn"
 VPC_NAT_IMAGE="vpc-nat-gateway"
 VERSION="v1.12.3"
 IMAGE_PULL_POLICY="IfNotPresent"
-POD_CIDR="10.16.0.0/16"                # Do NOT overlap with NODE/SVC/JOIN CIDR
-POD_GATEWAY="10.16.0.1"
-SVC_CIDR="10.96.0.0/12"                # Do NOT overlap with NODE/POD/JOIN CIDR
+POD_CIDR="10.244.0.0/16"                # Do NOT overlap with NODE/SVC/JOIN CIDR
+POD_GATEWAY="10.244.0.1"
+SVC_CIDR="10.96.0.0/16"                # Do NOT overlap with NODE/POD/JOIN CIDR
 JOIN_CIDR="100.64.0.0/16"              # Do NOT overlap with NODE/POD/SVC CIDR
 PINGER_EXTERNAL_ADDRESS="114.114.114.114"  # Pinger check external ip probe
 PINGER_EXTERNAL_DOMAIN="alauda.cn."         # Pinger check external domain probe
@@ -3557,7 +3557,12 @@ spec:
           image: "$REGISTRY/kube-ovn:$VERSION"
           imagePullPolicy: $IMAGE_PULL_POLICY
           command: 
-          - /kube-ovn/start-ovs.sh
+          - /bin/bash
+          - -c
+          - |
+              ln -sf /bin/true /usr/sbin/modinfo
+              ln -sf /bin/true /usr/sbin/modprobe
+              exec /kube-ovn/start-ovs.sh
           securityContext:
             runAsUser: 0
             privileged: true
@@ -4263,6 +4268,7 @@ spec:
           args:
           - --external-address=$PINGER_EXTERNAL_ADDRESS
           - --external-dns=$PINGER_EXTERNAL_DOMAIN
+          - --ds-namespace=cozy-kubeovn
           - --logtostderr=false
           - --alsologtostderr=true
           - --log_file=/var/log/kube-ovn/kube-ovn-pinger.log
